### PR TITLE
[Merged by Bors] - generate block right away in the non-optimistic case

### LINKED ISSUE #4482: block generator: don't enforce sequential generation for non-optimistic blocks
current implementation checks that the previous layer was applied before generating new block based on hare results. this is required for optimistic execution, as we actually apply the block before filtering out ineffective transactions.

however this may result in the network unable to make any progress, and apply empty layers indefinitely. to enter such state there should be a gap without any ballots, so that verified tortoise layer will be naturally delayed, from that point even if hare completes successfully - block generator won't be able to make any progress, and tortoise will vote against the layer.

example from recent network

latest ballot votes abstain on 1585
> 8aa757", "module": "proposalListener", "requestId": "b2fdde50-9833-4e89-a55c-a6b8690d3187", "proposal_id": "2c49360888", "ballot_id": "67c1ea3c32", "layer_id": 1586, "ballot_id": "67c1ea3c32", "layer_id": 1586, "epoch_id": 2, "smesher": "82aad33c2888a3c033ca9bc5329862f112d3728fc503830c41e3c7a0a1fdc2bb", "opinion hash": "85badb187c", "base_ballot": "3705bfc9aa", "support": 1, "against": 0, "abstain": 1, "atx_id": "5e859bad76", "ref_ballot": "d375a44105", "active_set_size": 0, "beacon": "00000000", "votes": {"base": "3705bfc9aa", "support": [{"id": "da7b242ad6", "layer": 1567, "height": 0}], "against": [], "abstain": ["1585"]}, "name": "proposalListener"}

executor applied 1583 as empty
> 9227:2023-06-06T10:05:11.719+0200	INFO	9eba1.executor	executed block	{"node_id": "9eba1a2b30d76d51f0ca71fbc2dece8352aa62ed7e97a3f0a13c66bf588aa757", "module": "executor", "protocol": "hr1_outputloop", "sessionId": "76e97847-51ea-46b6-b5fe-0e6df66f9d76", "layer_id": 1583, "block_id": "11571d9410", "event": true, "block": "11571d9410", "state_hash": "0x309dd3e265f4f946630b9de08a6ccffd54b0d1405938dccca07be16860bec924", "duration": "529.003µs", "count": 0, "name": "executor"}

hare reaches a consensus
> 2023-06-06T10:12:05.589+0200	INFO	9eba1.hare	hare terminated with success	{"node_id": "9eba1a2b30d76d51f0ca71fbc2dece8352aa62ed7e97a3f0a13c66bf588aa757", "module": "hare", "protocol": "hr1_outputloop", "sessionId": "ab0e329e-be02-4143-a7bb-0de587d1289f", "layer_id": 1586, "num_proposals": 8, "name": "hare"}

but generator will be stuck because of abstains on 1585

the desired behavior would be not to enter perpetual voting with abstain and create non-optimistic block immediately

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
Closes #4482

## Changes
- when optimistic filtering is OFF, generate a block right away regardless of layer order
- when hare output empty block, or a non-optimistic generated block, certify the hare output right away

note that block execution is still kept sequential

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by dshulyak** [comment]
> note that block execution is still kept sequential

this is enforced by mesh, because results from tortoise are sequantial. maybe we can drop this elsewhere, so that it doesn't do unnecessary things?

File: blocks/generator.go [hunk 1]
```diff
@@ -193,116 +193,149 @@ func (g *Generator) getProposals(pids []types.ProposalID) ([]*types.Proposal, er
...
-		if err != nil {
-			logger.With().Error("failed to generate block", log.Err(err))
-			failGenCnt.Inc()
-			return err
-		} else if err = g.msh.AddBlockWithTXs(ctx, block); err != nil {
-			logger.With().Error("failed to add block", log.Err(err))
-			failErrCnt.Inc()
-			return err
+func (g *Generator) processHareOutput(out hare.LayerOutput) (*blockGenInfo, error) {
+	md, err := g.preprocess(out)
```
Comment (dshulyak) [inline blocks/generator.go]: it would be easier to understand the code if preprocess was inlined

File: blocks/generator.go [hunk 2]
```diff
@@ -193,116 +193,149 @@ func (g *Generator) getProposals(pids []types.ProposalID) ([]*types.Proposal, er
...
+					TxIDs:      md.tids,
+				},
+			}
+			bi.block.Initialize()
 		}
-	} else {
-		emptyOutputCnt.Inc()
 	}
+	if md == nil || !md.optFilter {
+		if err = g.postProcess(out.Ctx, out.Layer, bi.block); err != nil {
```
Comment (dshulyak) [inline blocks/generator.go]: maybe `saveAndCertify` or just `certifiy` ?

**Review by dshulyak — APPROVED** [review]

**Comment by countvonzero** [comment]
> > note that block execution is still kept sequential
> 
> this is enforced by mesh, because results from tortoise are sequantial. maybe we can drop this elsewhere, so that it doesn't do unnecessary things?

there are 2 threads that could execute a block:
A. block generator (calling `ProcessLayerPerHareOutput`)
B. syncer (calling `ProcessLayer`)

in a steady state, case A will be ahead of tortoise verification results. B is just a backup in case the node get disconnected during hare consensus. the sequential execution here (in `processLayers`) only serves to ensure that blocks are executed in-situ sequentially. there is no reason to treat optimistic/non-optimistic differently here.

**Comment by dshulyak** [comment]
> in a steady state, case A will be ahead of tortoise verification results.

tortoise will output that layer is not verified but there is a hare output, and it will be enough for mesh to allow execution without persisting validity. during optimistic filtering something else happens, block is executed and `applied` field is saved to database, after that mesh calls tortoise and re-checks that target block is already applied

**Review by dshulyak — CHANGES_REQUESTED** [review]
actually this change still has the same problem. you need to notify tortoise to vote on the non-optimistic block before previous blocks are applied. otherwise it will be the same behavior as was seen

**Comment by countvonzero** [comment]
> actually this change still has the same problem. you need to notify tortoise to vote on the non-optimistic block before previous blocks are applied. otherwise it will be the same behavior as was seen

@dshulyak the code in this PR will call `AddBlockWithTXs()` as soon as hare terminates. that calls `trtl.OnBlock()`. i thought that is enough to make tortoise to vote on a block?

**Comment by dshulyak** [comment]
> @dshulyak the code in this PR will call AddBlockWithTXs() as soon as hare terminates. that calls trtl.OnBlock(). i thought that is enough to make tortoise to vote on a block?

OnBlock will notify tortoise that full data is available locally. tortoise will vote according to OnHareOutput

**Comment by countvonzero** [comment]
> > in a steady state, case A will be ahead of tortoise verification results.
> 
> tortoise will output that layer is not verified but there is a hare output, and it will be enough for mesh to allow execution without persisting validity. during optimistic filtering something else happens, block is executed and `applied` field is saved to database, after that mesh calls tortoise and re-checks that target block is already applied

in case A the optimistic case, at the time of execution, the block is not generated yet. tortoise has no way to include it in the results. 

the workflow for optimistic case is:
hare terminate -> txs get executed -> block gets generated -> trtl.OnBlock() -> `applied` field is saved -> mesh get results from trtl and ensure state consistent.

i don't' understand what need changed here.

**Comment by countvonzero** [comment]
> > @dshulyak the code in this PR will call AddBlockWithTXs() as soon as hare terminates. that calls trtl.OnBlock(). i thought that is enough to make tortoise to vote on a block?
> 
> OnBlock will notify tortoise that full data is available locally. tortoise will vote according to OnHareOutput

ok. thanks. i'll make that change.

**Comment by countvonzero** [comment]
@dshulyak please take another look.
the code is changed to such that for empty block ID and non-optimistic block tortoise is notified as soon as hare terminate.

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]
thanks, this should vote correctly

>  === CONT  TestPartition_50_50
    logger.go:130: 2023-06-08T13:12:30.367Z	ERROR	TestPartition_50_50	test failed	***"err": "stream err from client boot-1: rpc error: code = Canceled desc = grpc: the client connection is closing"***
    partition_test.go:164: 
        	Error Trace:	/build/systest/tests/partition_test.go:164
        	            				/build/systest/tests/partition_test.go:194
        	Error:      	Received unexpected error:
        	            	stream err from client boot-1: rpc error: code = Canceled desc = grpc: the client connection is closing
        	Test:       	TestPartition_50_50

i think pods in this test were interrupted prematurely, usually there are more layers
image
[grafana](https://grafana.spacemesh.dev/explore?orgId=1&left=%7B%22datasource%22:%22prometheus-gke-dev%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22spacemesh_tortoise_state_layers%7Bkubernetes_namespace%3D%5C%22test-ghgd%5C%22,%20kubernetes_pod_name%3D%5C%22boot-1-cdbc99d5b-khht5%5C%22%7D%22,%22range%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22prometheus-gke-dev%22%7D,%22editorMode%22:%22code%22%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22prometheus-gke-dev%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22spacemesh_tortoise_state_layers%7Bkubernetes_namespace%3D%5C%22test-ghgd%5C%22,%20kubernetes_pod_name%3D%5C%22boot-0-5dc88849b-42mhc%5C%22%7D%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221686228908387%22,%22to%22:%221686230296428%22%7D%7D)

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
TestPartition_50_50 consistently fails at layer 22 :(
will need more debugging to understand.

**Comment by dshulyak** [comment]
there might be a gap in events that are watched by a test, [tortoise](https://grafana.spacemesh.dev/explore?orgId=1&left=%7B%22datasource%22:%22loki-gke-dev%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22test-aeyj%5C%22,%20pod%3D%5C%22smesher-27-6dcc86c5b8-5rlcg%5C%22%7D%20%7C%3D%20%5C%22candidate%20layer%20is%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-gke-dev%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221686277895236%22,%22to%22:%221686280702016%22%7D%7D) and [mesh](https://grafana.spacemesh.dev/explore?orgId=1&left=%7B%22datasource%22:%22loki-gke-dev%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22test-aeyj%5C%22,%20pod%3D%5C%22smesher-27-6dcc86c5b8-5rlcg%5C%22%7D%20%7C~%20%5C%22consensus%20results%7Cexecutor%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-gke-dev%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221686277895236%22,%22to%22:%221686280702016%22%7D%7D) worked out correctly

maybe somehow mesh notifies about layers that are awaited by test before the actual test finishes

do you know if we watch this event in tests?
> events.ReportLayerUpdate(events.LayerUpdate{
		LayerID: layerID,
		Status:  events.LayerStatusTypeApproved,
	})

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [systest-status](https://github.com/spacemeshos/go-spacemesh/actions/runs/5230550454/jobs/9444347996)

test passed. failed due to an unknown kubernetes error
```
=== CONT  TestPartition_50_50
    logger.go:130: 2023-06-10T15:23:04.083Z	INFO	TestPartition_50_50	namespace was deleted	***"namespace": "test-rpoz"***
--- PASS: TestPartition_50_50 (925.24s)
=== CONT  TestFallback
    logger.go:130: 2023-06-10T15:23:04.135Z	INFO	TestFallback	using	***"namespace": "test-qzao"***
rpc error: code = NotFound desc = an error occurred when try to find container "5c50d9e8ccbdb0e30cf118da4776fb3ac0c8a2bca52ef8ab4848ce31aeaae6dd": not foundtest_job_name=systest-7f6b952-1686409567 ./wait_for_job.sh
Waiting for job/systest-7f6b952-1686409567 to complete or fail
```

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**